iroh/src/magicsock/transports.rs

```diff
@@ -0,0 +1,434 @@
+                    if transport.is_valid_send_addr(url, node_id) {
+                        match transport.poll_send(url.clone(), *node_id, transmit) {
+                            Poll::Pending => {}
+                            Poll::Ready(res) => return Poll::Ready(res),
+                        }
+                    }
+                }
+            }
+        }
+        Poll::Pending
+    }
+
+    /// Tries to recv data, on all available transports.
+    pub(crate) fn poll_recv(
+        &self,
+        cx: &mut Context,
+        bufs: &mut [IoSliceMut<'_>],
+        metas: &mut [quinn_udp::RecvMeta],
+        source_addrs: &mut [Addr],
+    ) -> Poll<io::Result<usize>> {
+        debug_assert_eq!(bufs.len(), metas.len(), "non matching bufs & metas");
+
+        // TODO: randomization
+        macro_rules! poll_transport {
+            ($socket:expr) => {
+                match $socket.poll_recv(cx, bufs, metas, source_addrs)? {
+                    Poll::Pending | Poll::Ready(0) => {}
+                    Poll::Ready(n) => {
+                        return Poll::Ready(Ok(n));
+                    }
+                }
+            };
+        }
+
+        #[cfg(not(wasm_browser))]
+        for transport in &self.ip {
+            poll_transport!(transport);
+        }
+        for transport in &self.relay {
+            poll_transport!(transport);
+        }
+        Poll::Pending
+    }
+
+    /// Returns a list of all currently known local addresses.
+    ///
+    /// For IP based transports this is the [`SocketAddr`] of the socket,
+    /// for relay transports, this is the home relay.
+    pub(crate) fn local_addrs(&self) -> Vec<Addr> {
+        self.local_addrs_watch().get().expect("not disconnected")
+    }
+
+    /// Watch for all currently known local addresses.
+    #[cfg(not(wasm_browser))]
+    pub(crate) fn local_addrs_watch(&self) -> impl Watcher<Value = Vec<Addr>> + Send + Sync {
+        let ips = watcher::Join::new(self.ip.iter().map(|t| t.local_addr_watch()));
+        let relays = watcher::Join::new(self.relay.iter().map(|t| t.local_addr_watch()));
+
+        (ips, relays)
+            .map(|(ips, relays)| {
+                ips.into_iter()
+                    .map(Addr::from)
+                    .chain(
+                        relays
+                            .into_iter()
+                            .flatten()
+                            .map(|(relay_url, node_id)| Addr::RelayUrl(relay_url, node_id)),
+                    )
+                    .collect()
+            })
+            .expect("disconnected")
+    }
+
+    #[cfg(wasm_browser)]
+    pub(crate) fn local_addrs_watch(&self) -> impl Watcher<Value = Vec<Addr>> + Send + Sync {
+        let relays = self.relay.iter().map(|t| {
+            t.local_addr_watch()
+                .map(move |t| t.map(|(url, id)| Addr::RelayUrl(url, id)))
+                .expect("disconnected")
+        });
+        watcher::JoinOpt::new(relays)
```

Comment:
Use `.map(|ls| ls.into_iter().flatten().collect())` instead? :upside_down_face: 
I'm still a firm believer in having as few (but powerful) combinators as possible.

Reply:
I am failing to implement this in a way that it compiles

Reply:
```rs
#[cfg(wasm_browser)]
pub(crate) fn local_addrs_watch(&self) -> impl Watcher<Value = Vec<Addr>> + Send + Sync {
    let relays = self.relay.iter().map(|t| t.local_addr_watch());
    watcher::Join::new(relays)
        .map(|relays| relays.into_iter().flatten().map(Addr::from).collect())
        .expect("disconnected")
}
```

Reply:
thanks